cmd/pop/main.go

```diff
@@ -0,0 +1,435 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azqueue"
+	"github.com/Azure/moby-packaging/pkg/queue"
+)
+
+const (
+	stagingAccountName = "moby"
+	prodAccountName    = "mobyreleases"
+	prodContainerName  = "moby"
+	queueName          = "moby-packaging-signing-and-publishing"
+)
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Fprintf(os.Stderr, "##vso[task.logissue type=error;]%s\n", err)
+		os.Exit(1)
+	}
+}
+
+type downloadArgs struct {
+	outDir       string
+	messagesFile string
+}
+
+type uploadArgs struct {
+	signedDir    string
+	messagesFile string
+}
+
+type fixupQueueArgs struct {
+	messagesFile string
+}
+
+type QueueMessageDeserialize struct {
+	Content         queue.Message `json:"content"`
+	DequeueCount    int           `json:"dequeueCount"`
+	ExpirationTime  string        `json:"expirationTime"`
+	ID              string        `json:"id"`
+	InsertionTime   string        `json:"insertionTime"`
+	PopReceipt      string        `json:"popReceipt"`
+	TimeNextVisible string        `json:"timeNextVisible"`
+}
+
+func (m *QueueMessageDeserialize) UnmarshalJSON(data []byte) error {
+	// we have to do this to keep from exploding the call stack
+	// see - https://medium.com/@turgon/json-in-go-is-magical-c5b71505a937
+
+	type Aux struct {
+		Content         string `json:"content"`
```

Comment:
Not sure I understand why this needs to be a string that we later unmarshal again?

Reply:
The Queue message format (defined by azure) allows for a single `content` string for data. If we give it the actual data structure we want (our message format), it won't unmarshal from the JSON response from the queue (it throws an error because we're trying to unmarshal a string field into a struct). So we unmarshal that as a string into a temporary struct, then unmarshal that string.

Happy to change if you know a better way, I just want to have access to both queue metadata (dequeue count, message id, pop receipt, etc) and our structured data about the packages (url, sha256 sum, the build spec).

Reply:
Here's an example of what I'm thinking based on what I see in the dequeued messages in the ADO job:

```go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

type Envelope struct {
	Content           string    `json:"content"`
	DequeueCount      int       `json:"dequeueCount"`
	InsertionTime     time.Time `json:"insertionTime"`
	ExpirationTime    time.Time `json:"expirationTime"`
	TimeNexstVisiable time.Time `json:"timeNextVisible"`
	ID                string    `json:"id"`
}

type Message struct {
	Artifact Artifact `json:"artifact"`
	// other fields
}

type Artifact struct {
	Name      string `json:"name"`
	URI       string `json:"uri"`
	Sha256Sum string `json:"sha256sum"`
	// other fields
}

func main() {
	dt, err := os.ReadFile("messages.json")
	if err != nil {
		panic(err)
	}

	var qm []Envelope
	err = json.Unmarshal(dt, &qm)
	if err != nil {
		panic(err)
	}
	fmt.Println(qm[0])

	var msg Message
	err = json.Unmarshal([]byte(qm[0].Content), &msg)
	if err != nil {
		panic(err)
	}
	fmt.Println(msg)
}
```

We aren't really interested in the full struct, at least not as it relates to the data inside of `content` so we can treat them as entirely separate entities and not have to do this dance with `MarshalJSON`.

I also think we should base64 encode our messages since these are getting stored as strings.
In which case unmarshalling it becomes this:


```go
var msg Message
dt, err = base64.StdEncoding.DecodeString(qm[0].Content)
if err != nil {
	panic(err)
}
err = json.Unmarshal(dt, &msg)
if err != nil {
	panic(err)
}
```

Reply:
Looks good, thanks for the detailed explanation

Reply:
done, LMK what you think of the changes. Most recent commit should make it clear. I'm going to test manually on a branch too

Reply:
tested, and the changes don't cause any obvious problems.

Reply:
tested, and the changes don't cause any obvious problems.